src/main/java/today/sesac/versebyverse/reaction/service/ReactionService.java

```diff
@@ -0,0 +1,210 @@
+import today.sesac.versebyverse.global.domain.Emotion;
+import today.sesac.versebyverse.member.entity.Member;
+import today.sesac.versebyverse.member.service.MemberService;
+import today.sesac.versebyverse.post.entity.Post;
+import today.sesac.versebyverse.post.repository.PostRepository;
+import today.sesac.versebyverse.reaction.dto.request.ReactionRequestDto;
+import today.sesac.versebyverse.reaction.dto.response.ReactionResponseDto;
+import today.sesac.versebyverse.reaction.entity.Reaction;
+import today.sesac.versebyverse.reaction.repository.ReactionRepository;
+import today.sesac.versebyverse.reaction.utils.TargetType;
+
+@Service
+@RequiredArgsConstructor
+@Transactional
+public class ReactionService {
+
+    private final ReactionRepository reactionRepository;
+    private final MemberService memberService;
+    private final PostRepository postRepository;
+    private final CommentRepository commentRepository;
+
+    /**
+     * 게시글과 댓글 반응 추가하기.
+     */
+    public ReactionResponseDto addReaction(ReactionRequestDto reactionRequestDto,
+                                           Long targetId, Long memberId, TargetType targetType) {
+
+        saveReaction(targetType, targetId, memberId, reactionRequestDto);
+
+        return addCountByReactionType(reactionRequestDto.getType(), targetId, targetType);
+    }
+
+    /**
+     * 게시글과 댓글 반응 삭제하기
+     *
+     * @param type       감정 타입
+     * @param targetId   댓글 or 게시글 id
+     * @param memberId   회원 id
+     * @param targetType 댓글 or 게시글
+     */
+    @Transactional
+    public ReactionResponseDto deleteReaction(String type, Long targetId, Long memberId,
+                                              TargetType targetType) {
+
+        //type을 대문자로 변환
+        Emotion emotion = Emotion.valueOf(type.toUpperCase());
+
+        //해당 postId에 회원이 눌렀던 감정을 삭제
+        deleteReactionIfExists(targetType, targetId, memberId, emotion);
+
+        return addCountByReactionType(emotion, targetId, targetType);
+    }
+
+    /**
+     * 게시글과 댓글 반응 수정하기
+     *
+     * @param reactionRequestDto 수정된 감정 타입
+     * @param targetType         수정된 게시글 or 댓글
+     * @param targetId           바뀐 게시글 or 댓글의 id
+     * @param memberId           회원 id
+     */
+    @Transactional
+    public ReactionResponseDto updateReaction(TargetType targetType, Long targetId,
+                                              ReactionRequestDto reactionRequestDto,
+                                              Long memberId) {
+
+        //1.회원이 기존에 눌렀던 감정 삭제
+        Optional<Reaction>
+                existingReaction = findExistingReaction(targetType, targetId, memberId);
+
+        if (existingReaction.isPresent()) {
+            reactionRepository.deleteById(existingReaction.get().getId());
+            reactionRepository.flush();
+        }
+
+        //2.해당 targetId에 memberId가 추가한 감정 삽입
+        saveReaction(targetType, targetId, memberId, reactionRequestDto);
```

Comment:
P2
구두 논의 내용.
entity에 update 메서드(이름은 고민 필요) 만들어서 변경 감지 기능을 이용하면 간단할 것 같습니다. 

```suggestion
reaction.updateReaction(reactionRequestDto.getType());
```

Reply:
[e936608](https://github.com/SproutJavaSesac/versebyverse-backend/pull/164/commits/e93660821c25076b9bd27c60ceb6d64c3a281f6d)